draft-ietf-masque-h3-datagram.md

```diff
@@ -117,28 +123,181 @@ following format (using the notation from the "Notational Conventions" section
 of {{QUIC}}):
 
 ~~~
-HTTP/3 DATAGRAM Frame {
-  Flow ID (i),
+HTTP/3 Datagram {
+  Quarter Stream ID (i),
+  Context ID (i),
   HTTP/3 Datagram Payload (..),
 }
 ~~~
 {: #h3-datagram-format title="HTTP/3 DATAGRAM Frame Format"}
 
-Flow ID:
+Quarter Stream ID:
 
-: A variable-length integer indicating the Flow ID of the datagram (see
-{{datagram-flows}}).
+: A variable-length integer that contains the value of the client-initiated
+bidirectional stream that this datagram is associated with, divided by four.
+(The division by four stems from the fact that HTTP requests are sent on
+client-initiated bidirectional streams, and those have stream IDs equal to zero
+modulo four.)
+
+Context ID:
+
+: A variable-length integer indicating the context ID of the datagram (see
+{{datagram-contexts}}).
 
 HTTP/3 Datagram Payload:
 
 : The payload of the datagram, whose semantics are defined by individual
 applications. Note that this field can be empty.
 
 Endpoints MUST treat receipt of a DATAGRAM frame whose payload is too short to
-parse the flow ID as an HTTP/3 connection error of type
+parse the Quarter Stream ID or Context ID as an HTTP/3 connection error of type
 H3_GENERAL_PROTOCOL_ERROR.
```

Comment:
This requirement seems awkard to implement as specified. I think some of that comes from mixing length checks and parsing. To elabortate:


* it states endpoints, but what about intermediaries (they need to parse just the Quarter Stream ID but should they forward a frame whose length is exactly the Quarter Stream ID?)
* for an endpoint parsing both fields it probably needs to be an 'and' 
* since the fields are varints its harder to check the DATAGRAM payload length can satisfy the field requirements. The obvious case is if payload length is 1 (two minimally encoded varints) but other combos make it harder.

The way that H3 specs the requirement isn't a perfect fit but I think the spirit is more closely align with what you're trying to say - https://tools.ietf.org/html/draft-ietf-quic-http-34#section-7.1

Reply:
Good point, one of those is hop-by-hop but the other is end-to-end. I've tweaked the requirements here to reflect that.